File: backend/app/schemas/report.py
from datetime import datetime
from typing import Literal

from pydantic import BaseModel, Field, field_validator

from app.schemas.location import LocationDetail, LocationInput, LocationSummary

DamageLevel = Literal["minimal", "partial", "complete"]
InfraType = Literal[
    "residential",
    "commercial",
    "government",
    "utility",
    "transport",
    "community",
    "public_space",
    "other",
]
ReportStatus = Literal["pending", "validated", "rejected"]
SubmissionChannel = Literal["app", "whatsapp", "web", "sms"]


class ReportCreate(BaseModel):
    crisis_id: str
    damage_level: DamageLevel
    infra_type: InfraType
    infra_subtype: str | None = Field(default=None, max_length=100)
    infra_name: str | None = Field(default=None, max_length=200)
    debris_present: bool
    nature_of_crisis: str | None = Field(default=None, max_length=50)
    description_raw: str | None = None
    reporter_name: str | None = Field(default=None, max_length=100)
    source_language: str | None = Field(default=None, max_length=10)
    submission_channel: SubmissionChannel = "app"
    collected_at: datetime
    location: LocationInput

    @field_validator("reporter_name", mode="before")
    @classmethod
    def normalize_reporter_name(cls, value: str | None) -> str:
        if value is None or not str(value).strip():
            return "anonymous"
        return str(value).strip()


class ReportStatusUpdate(BaseModel):
    status: ReportStatus


class ReportOut(BaseModel):
    id: str
    crisis_id: str
    location_id: str
    damage_level: DamageLevel
    infra_type: InfraType
    infra_subtype: str | None = None
    infra_name: str | None = None
    debris_present: bool
    nature_of_crisis: str | None = None
    description_raw: str | None = None
    description_translated: str | None = None
    reporter_name: str = "anonymous"
    source_language: str | None = None
    is_latest_version: bool
    version_number: int
    submission_channel: SubmissionChannel
    status: ReportStatus
    collected_at: datetime
    submitted_at: datetime
    location: LocationSummary | LocationDetail | None = None


class ReportVersionOut(BaseModel):
    id: str
    version_number: int
    damage_level: DamageLevel
    is_latest_version: bool
    collected_at: datetime
    submitted_at: datetime


class CrisisReportsQuery(BaseModel):
    damage_level: DamageLevel | None = None
    infra_type: InfraType | None = None
    admin_level_2: str | None = None
    status: ReportStatus | None = None
    date_from: datetime | None = None
    date_to: datetime | None = None
    page: int = Field(default=1, ge=1)
    limit: int = Field(default=50, ge=1, le=200)

File: frontend/src/types/report.ts
export type CrisisType = "natural_hazard" | "technological" | "human_made";
export type CrisisStatus = "active" | "closed";

export interface Crisis {
  id: string;
  name: string;
  crisis_type: CrisisType;
  crisis_subtype: string;
  epicenter_lat?: number | null;
  epicenter_lng?: number | null;
  status: CrisisStatus;
  is_unlisted?: boolean;
  form_template_id?: string | null;
  onset_at: string;
  created_at: string;
}

export interface ReportingOptions {
  crises: Crisis[];
  unlisted_crisis_id: string;
  nearest_crisis_id: string | null;
}

export type DamageLevel = "minimal" | "partial" | "complete";

export type InfraType =
  | "residential"
  | "commercial"
  | "government"
  | "utility"
  | "transport"
  | "community"
  | "public_space"
  | "other";

export type LocationMethod = "gps" | "what3words" | "manual" | "exif";

export interface ReportLocationInput {
  latitude?: number;
  longitude?: number;
  what3words?: string;
  location_method: LocationMethod;
  building_footprint_id?: string;
}

export interface ReportCreateInput {
  crisis_id: string;
  damage_level: DamageLevel;
  infra_type: InfraType;
  infra_subtype?: string;
  infra_name?: string;
  debris_present: boolean;
  nature_of_crisis?: string;
  description_raw?: string;
  reporter_name?: string;
  source_language?: string;
  submission_channel: "mobile" | "web";
  collected_at: string;
  location: ReportLocationInput;
  form_responses?: Record<string, unknown> | null;
}

export interface ReportLocation {
  id: string;
  latitude: number;
  longitude: number;
  what3words?: string | null;
  admin_level_1?: string | null;
  admin_level_2?: string | null;
  admin_level_3?: string | null;
}

export interface Report {
  id: string;
  crisis_id: string;
  location_id: string;
  damage_level: DamageLevel;
  infra_type: InfraType;
  infra_subtype?: string | null;
  infra_name?: string | null;
  debris_present: boolean;
  nature_of_crisis?: string | null;
  description_raw?: string | null;
  description_translated?: string | null;
  reporter_name?: string;
  source_language?: string | null;
  is_latest_version: boolean;
  version_number: number;
  submission_channel: "mobile" | "web";
  status: string;
  collected_at: string;
  submitted_at: string;
  location?: ReportLocation;
  form_responses?: Record<string, unknown> | null;
}

export interface ApiEnvelope<T> {
  data: T | null;
  error: { code: string; message: string } | null;
}

export interface MapReportProperties {
  location_id: string;
  report_id: string;
  damage_level: DamageLevel;
  infra_type: InfraType;
  nature_of_crisis?: string | null;
  report_count: number;
  admin_level_2?: string | null;
  latest_photo_thumbnail?: string | null;
}

export interface MapGeoJsonFeature {
  type: "Feature";
  geometry: {
    type: "Point";
    coordinates: [number, number];
  };
  properties: MapReportProperties;
}

export interface MapFeatureCollection {
  type: "FeatureCollection";
  features: MapGeoJsonFeature[];
  total: number;
}

export interface MapReportPin {
  id: string;
  locationId: string;
  latitude: number;
  longitude: number;
  damageLevel: DamageLevel;
  infraType: InfraType;
  natureOfCrisis?: string | null;
  reportCount: number;
  adminLevel2?: string | null;
  thumbnail?: string | null;
}

export type PhotoMimeType = "image/jpeg" | "image/png" | "image/webp";

export interface PhotoInitiateInput {
  mime_type: PhotoMimeType;
  file_size_kb: number;
}

export interface PhotoInitiateResult {
  photo_id: string;
  storage_path: string;
  upload_url: string;
  expires_in: number;
}

export interface PhotoConfirmInput {
  photo_id: string;
  storage_path: string;
  file_size_kb: number;
  mime_type: PhotoMimeType;
  captured_at?: string;
  gps_lat?: number;
  gps_lng?: number;
}

export interface Photo {
  id: string;
  report_id: string;
  storage_path: string;
  signed_url?: string | null;
  thumbnail_url?: string | null;
  file_size_kb?: number | null;
  mime_type?: string | null;
  captured_at?: string | null;
  uploaded_at: string;
}

export interface ReportVersion {
  id: string;
  version_number: number;
  damage_level: DamageLevel;
  is_latest_version: boolean;
  collected_at: string;
  submitted_at: string;
}

export interface ReportDetail extends Report {
  photos: Photo[];
}
